### DISCUSSION TITLE
search output library

### DISCUSSION THREAD

**Discussion by JM-Bader** [vdemichev/DiaNN#444] [Q&A] [answered]
One can create a new spectral library as part of a DIANN search. 

Q1: What is the envisioned application case of that?
Q2: If the search was started with an input library and produces an output library, how do they differ? Does the new library only contain the precursors identified in the search?
Q3: Sometimes directDIA (ie. library-free DIA) can be slow. Is it legitimate to run such a library-free search on a subset of samples to generate a library and use this library in a second search on the full set of samples ? 
Q4: DIANN has the Speed and RAM usage options which are kind of an alternative to the suggested approach in Q3. Q3 should be better than compromising via these settings, isn't it?

**Comment by vdemichev** [comment, marked as answer]
Q1: (i) creation of libraries from DIA data, (ii) this is done automatically during MBR.
Q2: yes, only precursors at 1% global FDR, or whatever FDR filter is set; what info is stored for these precursors is determined by the 'Library generation' setting.
Q3: Yes, just make sure the subset of samples used is representative of the whole dataset.
Q4: Running on a subset should have no performance penalty if you use it on hundreds of runs. Amoung speed options only ultra-fast is a game changer, but it significantly reduces IDs and inflates FDR a bit, so not recommended.

**Reply by JM-Bader** [reply]
thank you for the feedback. Very clear. 

I ran searches as proposed in Q3 on a subset of only 96 samples that are representative of the larger cohort (thousands of samples).

Q5: This initial search was performed with a library (which is very large and thus the search takes too long for the entire larger cohort).
As I control I took the outputted library and ran the same 96 samples again. in this re-search, proteome depth was reduced about 15%. Why is that ? 

Q6: for comparison, I did the same thing with a library-free initial search on 96 samples. I used DIANN's FASTA digest, Deep learning-based prediction, heuristic protein inference and protein inference dropdown set to genes. The output pg.matrix does have the protein grouping that I want. However, when using the library that was generated in this initial search for a Re-Search with FASTA digest off, Deep learning off,  Reannotate off, Protein inference off, the resulting output still has the non-heuristic protein grouping (with much higher numbers of protein groups). 
Why is that ? 
How to fix it if I want the heuristic protein grouping ?

**Reply by vdemichev** [reply]
Q5: strange. Can you please share the logs?
Q6: Libraries generated from DIA data with DIA-NN's lib-free search must be used with protein inference enabled. Otherwise there's just complete lack of protein grouping.

**Reply by JM-Bader** [reply]
Re Q5: 
initial search: 
[DA-4O1.1_PL01_DA-SER1.19_C15Lib05_.log.txt](https://github.com/vdemichev/DiaNN/files/9183041/DA-4O1.1_PL01_DA-SER1.19_C15Lib05_.log.txt)

re-search with library created from original search: 
[DA-4O1.3_PL01_DA-SER1.19_C15Lib05_optRes_RE-search_.log.txt](https://github.com/vdemichev/DiaNN/files/9183046/DA-4O1.3_PL01_DA-SER1.19_C15Lib05_optRes_RE-search_.log.txt)

**Reply by JM-Bader** [reply]
Re Q6: is there a way to use the library generated from a library-free search to yield final results that have the heuristic protein grouping?

**Reply by vdemichev** [reply]
Some runs gain from re-search, some lose a bit. Overall seems about the same? Not gain is expected, if you are starting with a library that is quite project-specific already.

Q6: Just enable protein inference and heuristic protein grouping, and do specify a FASTA database.

**Reply by JM-Bader** [reply]
Re Q5: This is the quantitation of protein groups across raw files
![image](https://user-images.githubusercontent.com/98888664/180829404-1bec940c-7217-40fe-864a-960ef65c9099.png)
sure, no gain is expected. But the drop in identifications is not expected either.

**Reply by vdemichev** [reply]
Just noticed, your first-pass analysis (no research) already uses MBR. So the end result is as good as it gets. The 'research' also uses MBR - but here it must be off, as the newly created DIA-based lib is super project-specific. So what you observe is fully expected.

**Reply by vdemichev** [reply]
Check the re-search run, 'first-pass' output

**Reply by JM-Bader** [reply]
thanks for the hint. I will. 
However, the first pass result does not have maxLFQ normalisation, correct ?
Would it help in my case to run the re-search again without MBR ? 

Apart from normalisation by MaxLFQ, is there another difference between first and second pass ? (even if MBR is disabled).
Or in other words, can I further save time by using first pass results for final study if MBR is disabled because the library is so project-specific?

**Reply by JM-Bader** [reply]
RE Q5: 
I tried it (Re-Search and taking first pass results, Re-Search with MBR disabled), however the ID drop persists: 
![image](https://user-images.githubusercontent.com/98888664/181058179-ffa0ba9e-4638-47b1-831d-35b910ea0fba.png)

So there should be another reason for the performance drop
What could it be?

**Reply by vdemichev** [reply]
How do logs look like?

**Reply by vdemichev** [reply]
First pass does have MaxLFQ. First pass should only be used if you have highly project-specific spectral library. In all other cases - MBR & second pass.

**Reply by JM-Bader** [reply]
initial search: 
[DA-4O3.1_PL01_DA-SER1.19_C17Lib02_optRes_.log.txt](https://github.com/vdemichev/DiaNN/files/9191788/DA-4O3.1_PL01_DA-SER1.19_C17Lib02_optRes_.log.txt)

Re-Search with MBR: 
[DA-4O3.3_PL01_DA-SER1.19_C17Lib02_optRes_RE-search_.log.txt](https://github.com/vdemichev/DiaNN/files/9191790/DA-4O3.3_PL01_DA-SER1.19_C17Lib02_optRes_RE-search_.log.txt)

Re-Search without MBR:
[DA-4O3.4_PL01_DA-SER1.19_C17Lib02_optRes_RE-search_noMBR_.log.txt](https://github.com/vdemichev/DiaNN/files/9191794/DA-4O3.4_PL01_DA-SER1.19_C17Lib02_optRes_RE-search_noMBR_.log.txt)

**Reply by vdemichev** [reply]
So? All as expected. "initial search:" - this is how it's supposed to be done. Re-Search without MBR: - effectively the same, but done manually. Re-Search with MBR: - first pass is the same as no MBR, the MBR result (second pass) is worse. Something is wrong with the filtering used by the script maybe? You can see that initial search has 1017 unique proteins for the first run, Re-Search without MBR has 1025. For all other runs should also be about the same numbers.

**Reply by vdemichev** [reply]
Protein inference is turned off in all cases - only OK if the library already has protein groups assembled. Sure this is the case?

**Reply by JM-Bader** [reply]
> So? All as expected. "initial search:" - this is how it's supposed to be done. Re-Search without MBR: - effectively the same, but done manually. Re-Search with MBR: - first pass is the same as no MBR, the MBR result (second pass) is worse. Something is wrong with the filtering used by the script maybe? You can see that initial search has 1017 unique proteins for the first run, Re-Search without MBR has 1025. For all other runs should also be about the same numbers.

Q7: "Something is wrong with the filtering used by the script maybe?".  I am not completely sure what you mean. I don't filter the library but take the library generated by the initial search. To quantify the search results (protein depth plotted), I use a loop that treats all pg.matrix files the same. Moreover, the script does not filter.  
I count valid values .agg(pd.notnull).agg(sum, axis='rows') across protein groups for each raw file. 
So I don't think there is a mistake/bias there. 

However, I find that there is a discrepancy between the number of proteins reported in the log and the actual numbers in the pg.matrix file:
intial search log:  1017 proteins
![image](https://user-images.githubusercontent.com/98888664/181071354-cdef15c4-e9f2-4809-96b1-d5facd88819f.png)
initial search actual results in pg.matrix: 1360  
![image](https://user-images.githubusercontent.com/98888664/181071852-7cb0a6b3-8e68-4840-9dc1-5be394312ee0.png)
For background: dict_PGmatrix['DA_4O3_1'] is the relevant pg.matrix of that search

**Reply by JM-Bader** [reply]
> Protein inference is turned off in all cases - only OK if the library already has protein groups assembled. Sure this is the case?

inital search: done with MSFragger lib and protein grouping from that. 
Re-searches: done with output lib from initial search. 
So I presume it is always the same grouping. Correct?

Secondly, my assumption is that protein inference generally is only needed for library-free searches. 

Thirdly, I am still confused about Reannotate (button),  Protein Inference (dropdown menu), heuristic protein inference (button, = --relaxed prot inf ?).  Wouldn't reannotate change the protein assembly as well?

**Reply by vdemichev** [reply]
I meant something might be wrong with how the script used to visualise ID numbers filters the final report.
Yes, log is different as different things (unique proteins vs protein groups) filtered differently are counted.
pg_matrix when researching will be filtered too strictly, need to use the main report instead. This explains it.
"inital search: done with MSFragger lib and protein grouping from that." - strange that the number of protein groups is higher than the number of isoforms...
Protein inference is always needed if the UniprotIDs column in the library does not contain already inferred proteins.
Reannotate simply fully replaces all the protein information in the library, by matching the peptides to the FASTA database provided.
Heuristic inference - yes, it just activates --relaxed-prot-inf.

**Reply by JM-Bader** [reply]
ok, thanks. 
You said the explanation is that pg.matrix is filtered more stringently in the research. Great that you found the reason. 

Q1: Is there a general discrepancy between pg.matrix and "main" .tsv ?
Q2: But I presume this should not be specific to the re-search? 
However, I also took the pg.matrix for the initial search > So the effect  of lower IDs in pg.matrix appears to be at least stronger for the Re-Search than for the initial search. Why is that?

**Reply by vdemichev** [reply]
There's no discrepancy, just the way pg_matrix is generated depends on whether MBR is used: with MBR it filters based on 'Lib.' q-values instead of 'Global.' q-values.

**Reply by JM-Bader** [reply]
Sorry, I don't really understand that yet. 
No matter whether MBR is used or not in the re-search, I get lower IDs in the Re-Search. So in both cases, using 'Lib' q-values or 'global' q-values, there are lower IDs just in the Re-Search

**Reply by vdemichev** [reply]
Can you please repeat the comparison using the main report? Based on what you showed previously, the numbers you plotted are fully expected.

**Reply by JM-Bader** [reply]
> I meant something might be wrong with how the script used to visualise ID numbers filters the final report.

Should the final (main) report be filtered, when I want protein-level information (basically pg.matrix information) ?

**Reply by JM-Bader** [reply]
> Can you please repeat a comparison using the main report? Based on what you showed previously, the nubmers you plotted are fully expected.

Yes, will do. I understand that this is the proposed solution. I just dont fully understand the context of it yet.

**Reply by vdemichev** [reply]
When searching with a DIA-based lib, that's what you do during research, regardless MBR or not, the numbers in pg_matrix are SEVERELY (in case of your runs) reduced because of too strinct filtering.

**Reply by vdemichev** [reply]
This is by design

**Reply by JM-Bader** [reply]
it works: there is no depth loss when using the first pass results from the main .tsv of the Re-Search (compared to the second pass results of the initial search): 
![image](https://user-images.githubusercontent.com/98888664/181274997-eb9545dd-77b7-47f2-97db-59de489684d5.png)
